community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/CompiledExecutionResult.scala

```diff
@@ -74,9 +74,19 @@ abstract class CompiledExecutionResult extends InternalExecutionResult {
 
   override def queryStatistics(): InternalQueryStatistics = InternalQueryStatistics()
 
+  private var successful = false
+  protected def success(): Unit = {
+    successful = true
+  }
+
   override def close(): Unit = {
-    ensureIterator()
-    innerIterator.close()
+    if (innerIterator != null) {
+      innerIterator.close( )
+    }
+    if (statement != null) {
```

Comment:
`innerIterator` I didn't change (I just made the close method not have to materialise the iterator, because having to exhaust all the data for close not only was unnecessary but lead to infinite recursion when making sure that exhaustion calls close).

Statement I feel is easier to have nullable than to have as an option, since it comes from generated java code. But if we could have a constructor that converts the null to an option field, that would be useful, unless we have to write that as an `apply`-method, because extending that from a java class is horrible...